python/artm/regularizers.py

```diff
@@ -0,0 +1,421 @@
+import uuid
+import random
+
+import messages_pb2
+import library as lib
+
+
+__all__ = [
+    'SmoothSparsePhiRegularizer',
+    'SmoothSparseThetaRegularizer',
+    'DecorrelatorPhiRegularizer',
+    'LabelRegularizationPhiRegularizer',
+    'SpecifiedSparsePhiRegularizer',
+    'ImproveCoherencePhiRegularizer'
+]
+
+CREATE_REGULARIZER_CONFIG = {
+    lib.RegularizerConfig_Type_SmoothSparsePhi: messages_pb2.SmoothSparsePhiConfig,
+    lib.RegularizerConfig_Type_SmoothSparseTheta: messages_pb2.SmoothSparseThetaConfig,
+    lib.RegularizerConfig_Type_DecorrelatorPhi: messages_pb2.DecorrelatorPhiConfig,
+    lib.RegularizerConfig_Type_LabelRegularizationPhi: messages_pb2.LabelRegularizationPhiConfig,
+    lib.RegularizerConfig_Type_SpecifiedSparsePhi: messages_pb2.SpecifiedSparsePhiConfig,
+    lib.RegularizerConfig_Type_ImproveCoherencePhi: messages_pb2.ImproveCoherencePhiConfig
+}
+
+
+def _reconfigure_field(obj, field, field_name, proto_field_name=None):
+    if proto_field_name is None:
+        proto_field_name = field_name
+    setattr(obj, '_' + field_name, field)
+    config = CREATE_REGULARIZER_CONFIG[obj._type]()
+    config.CopyFrom(obj._config)
+    if isinstance(field, list):
+        config.ClearField(proto_field_name)
+        for value in field:
+            getattr(config, proto_field_name).append(value)
+    else:
+        setattr(config, proto_field_name, field)
+    obj.regularizer.Reconfigure(obj.regularizer.config_.type, config)
+
+
+class Regularizers(object):
+    """Regularizers represents a storage of regularizers in ArtmModel (private class)
+
+    Args:
+      master (reference): reference to MasterComponent object, no default
+    """
+
+    def __init__(self, master):
+        self._data = {}
+        self._master = master
+
+    def add(self, config):
+        """Regularizers.add() --- add regularizer into ArtmModel
+
+        Args:
+          config (reference): reference to an object of ***Regularizer class, no default
+        """
+        if config.name in self._data:
+            raise ValueError('Regularizer with name ' + str(config.name) + ' is already exist')
+        else:
+            regularizer = self._master.CreateRegularizer(config.name, config.type, config.config)
+            config._regularizer = regularizer
+            self._data[config.name] = config
+
+    def __getitem__(self, name):
+        """Regularizers.__getitem__() --- get regularizer with given name
+
+        Args:
+          name (str): name of the regularizer, no default
+        """
+        if name in self._data:
+            return self._data[name]
+        else:
+            raise KeyError('No regularizer with name ' + name)
+
+    @property
+    def data(self):
+        return self._data
+
+
+class BaseRegularizer(object):
```

Comment:
- each field is duplicated: object field and field of the protobuf message
- lot of meaningless properties, copied many times 

... :(

I want a bit sophisticated base class:
```python
class BaseRegularizer(object):

    _TYPE = None
    _CONFIG_MESSAGE = None

    def __init__(self, name, tau, topic_names, **kwargs):
        if _TYPE is None or _CONFIG_MESSAGE is None:
            raise NotImplementedError('{} is abstract'.format(self.__class__))

        self.name = name if name is not None else '{}: {}'.format(self._TYPE, uuid.uuid1().urn)
        self.tau = tau

        self.config = self._CONFIG_MESSAGE()
        for field_name, value in kwargs.iteritems():
            setattr(self.config, field_name, value)
            # TODO: handle complex fields, use the self._FIELDS description

    def __getattr__(self, attr_name):
        if hasattr(self.config, attr_name):
            return getattr(self.config, attr_name)
        else:
            raise AttributeError(attr_name)

    def __setattr__(self, attr_name, value):
        if hasattr(self.config, attr_name):
            setattr(self.config, attr_name, value)
            self.lib_.Reconfigure(config)
        else:
            raise AttributeError(attr_name)
```

And a very concise and clean regularizer classes, like this:
```
class SmoothSparsePhiRegularizer(BaseRegularizer):
    """SmoothSparsePhiRegularizer is a regularizer in ArtmModel (public class)
    """

    _TYPE = lib.RegularizerConfig_Type_SmoothSparsePhi
    _CONFIG_MESSAGE = messages_pb2.SmoothSparsePhiConfig
    _FIELDS = [
        'topics',
        'namespaces',
        'dictionary_name',
    ]
```

or maybe with django-styled fields:
```python
class SmoothSparsePhiRegularizer(BaseRegularizer):
    """SmoothSparsePhiRegularizer is a regularizer in ArtmModel (public class)
    """

    _TYPE = lib.RegularizerConfig_Type_SmoothSparsePhi
    _CONFIG_MESSAGE = messages_pb2.SmoothSparsePhiConfig
    
    topics = fields.StringListFields()
    namespaces = fields.StringListFields()
    dictionary_name = fields.PathField()

```